libs/dmlf/src/basic_vm_engine.cpp

```diff
@@ -389,117 +196,267 @@ bool BasicVmEngine::HasState(std::string const &name) const
+                                 "Error(" + id + ") in output after running. " + errorMessage},
+                           console};
+  };
+
+  LedgerVariant output;
+  if (vmVariant.type_id <= vm::TypeIds::PrimitiveMaxId)
   {
-    return VmVariant(ledgerVariant.As<float>(), typeId);
+    switch (vmVariant.type_id)
+    {
+    case fetch::vm::TypeIds::Bool:
+    {
+      output = vmVariant.Get<bool>();
+      break;
+    }
+    case fetch::vm::TypeIds::Int8:
+    case fetch::vm::TypeIds::UInt8:
+    case fetch::vm::TypeIds::Int16:
+    case fetch::vm::TypeIds::UInt16:
+    case fetch::vm::TypeIds::Int32:
+    case fetch::vm::TypeIds::UInt32:
+    case fetch::vm::TypeIds::Int64:
+    {
+      output = vmVariant.Get<int>();
+      break;
+    }
+    case fetch::vm::TypeIds::Float32:
+    {
+      output = vmVariant.Get<float>();
+      break;
+    }
+    case fetch::vm::TypeIds::Float64:
+    {
+      output = vmVariant.Get<double>();
+      break;
+    }
+    case fetch::vm::TypeIds::Fixed32:
+    {
+      output = vmVariant.Get<fp32_t>();
+      break;
+    }
+    case fetch::vm::TypeIds::Fixed64:
+    {
+      output = vmVariant.Get<fp64_t>();
+      break;
+    }
+    case vm::TypeIds::Void:
+    case vm::TypeIds::Unknown:
+    {
+      break;
+    }
+    default:
+      return serializationError("Could not transform primitive type " +
+                                vm.GetTypeName(vmVariant.type_id));
+    }
   }
-  case fetch::vm::TypeIds::Float64:
+  else if (vmVariant.type_id == vm::TypeIds::String)
   {
-    return VmVariant(ledgerVariant.As<double>(), typeId);
+    output = vmVariant.Get<vm::Ptr<vm::String>>()->str;
   }
-  case fetch::vm::TypeIds::Fixed32:
-  case fetch::vm::TypeIds::Fixed64:
+  else
   {
-    return VmVariant(ledgerVariant.As<fp64_t>(), typeId);
-  }
-  default:
-    return VmVariant();
+    ExecutionContext executionContext(&vm, exec);
+
+    auto inside = vmVariant.Get<vm::Ptr<vm::Object>>();
+
+    serializers::MsgPackSerializer serializer;
+    try
+    {
+      inside->SerializeTo(serializer);
+      serializer.seek(0);
+    }
+    catch (std::exception &ex)
+    {
+      return serializationError("Serializing output threw error " + std::string(ex.what()));
+    }
+    catch (...)
+    {
+      return serializationError("Serializing output threw error. No details.");
+    }
+
+    try
+    {
+      serializer >> output;
+    }
+    catch (std::exception &ex)
+    {
+      return serializationError(" Deserializing output after running. Threw error " +
+                                std::string(ex.what()));
+    }
+    catch (...)
+    {
+      return serializationError(" Deserializing output after running. No details.");
+    }
+
+    if (output.IsArray())  // Convert inner type from int to fixedpoint if necessary
+    {
+      auto GetInnermostTypeId = [&vm](LedgerVariant const &ledgerVariant,
+                                      VmVariant const &    vmVariant) {
```

Comment:
The opening brace should probably be placed in the next line, to be in-line with other formatting rules.